cmd/k8s-operator/ports.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand/v2"
+	"sort"
+	"strconv"
+	"strings"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"tailscale.com/kube/kubetypes"
+)
+
+const (
+	portMax            = 65535
+	portMin            = 1024
+	testSvcName     = "test-node-port-range"
+
+	invalidSvcNodePort = 777777
+)
+
+var kubePortRangeErr = fmt.Sprintf("Service \"%s\" is invalid: spec.ports[0].nodePort: Invalid value: %d: provided port is not in the valid range. The range of valid ports is ", invalidSvcName, invalidSvcNodePort)
+
+type portRange struct {
+	Start  int
+	End    int
+	String string
+}
+
+func validateRange(s int, e int) error {
+	if s < 1 || s > 65535 {
+		return fmt.Errorf("invalid port value: %q", s)
+	}
+	if e < 1 || e > 65535 {
+		return fmt.Errorf("invalid port value: %q", e)
+	}
+	if s > e {
+		return fmt.Errorf("invalid port range: '%d-%d'", s, e)
+	}
+
+	return nil
+}
+
+// getKubeRange is a hacky function that attempts to create a deliberately invalid
+// Service with a NodePort that is too large. This gives us back an error message
+// which contains the configured NodePort Range set on the cluster.
+// We don't return an error here, if this validation doesn't work we don't perform it
+func getKubeRange(ctx context.Context, c client.Client, tsNamespace string) *portRange {
```

Comment:
Suggestion:
```suggestion
// getServicesNodePortRange is a hacky function that attempts to determine Service NodePort range by attempting to
// create a deliberately invalid Service with a NodePort that is too large and parses the returned validation error // to determine the port range. Returns nil if unable to determine port range.
// https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
func getServicesNodePortRange(ctx context.Context, c client.Client, tsNamespace string) *portRange {
```

Generally, we should also aim to finish sentences with a `.`